docs/android-faqs.md

```diff
@@ -0,0 +1,119 @@
+
+# Rust + Android FAQs
+
+### How do I expose Rust code to Kotlin?
+
+Use [UniFFI](https://mozilla.github.io/uniffi-rs/), which can produce Kotlin
+bindings for your Rust code from an interface definition file.
+
+If UniFFI doesn't currently meet your needs, please [open an issue](
+https://github.com/mozilla/uniffi-rs/issues) to discuss how the tool can
+be improved.
+
+As a last resort, you can make hand-written bindings from Rust to Kotlin,
+essentially manually performing the steps that UniFFI tries to automate
+for you:
+
+* Expose some `pub extern "C"` functions from your Rust code.  Use the
+  [ffi_support](https://docs.rs/ffi-support/0.1.3/ffi_support/) crate to help make
+  this easier, and consult the crate's documentation for tips and gotchas.
+* Write some low-level Kotlin bindings for the exposed functions, using
+  [JNA](https://github.com/java-native-access/jna) to load the compiled rust
+  code via shared library. Our [android support package](../components/support/android)
+  has some helpers to make this easier.
+* Write some higher-level Kotlin code that wraps those bindings in a nice safe
+  ergonomic Kotlin API, trying to mirror the structure of the original Rust code
+  as much as possible.
+
+### How should I name the package?
+
+Published packages should be named `org.mozilla.appservices.$NAME` where `$NAME`
+is the name of your component, such as `logins`.  The Java namespace in which
+your package defines its classes etc should be `mozilla.appservices.$NAME.*`.
+
+### How do I publish the resulting package?
+
+Add it to `.buildconfig-android.yml` in the root of this repository.
+This will cause it to be automatically included as part of our release
+publishing pipeline.
+
+### How do I know what library name to load to access the compiled rust code?
+
+Assuming that you're building the Rust code as part of the application-services
+build and release process, your `pub extern "C"` API should always be available
+from a file named `libmegazord.so`.
+
+### What challenges exist when calling back into Kotlin from Rust?
+
+There are a number of them. The issue boils down to the fact that you need to be
+completely certain that a JVM is associated with a given thread in order to call
+java code on it. The difficulty is that the JVM can GC its threads and will not
+let rust know about it. JNA can work around this for us to some extent, however
+there are difficulties.
+
+The approach it takes is essentially to spawn a thread for each callback
+invocation. If you are certain you’re going to do a lot of callbacks and they
+all originate on the same thread, you can tell it to cache these.
+
+Calling back from Rust into Kotlin isn’t too bad so long as you ensure the
```

Comment:
I don't quite get this para - the ones above basically say it is difficult because the threads might be GCs, but this says it's not bad so long as a you can avoid the GC without guidance on how to do this.